### PR TITLE
Create personal realm for newly registered users

### PR DESCRIPTION
**Pull request by habdelra**
This PR introduces the logic to create a realm for a newly registered user. Please make sure that you have registered all the matrix realm users before trying this out. I've also update our `pnpm drop-all-dbs` script so that we also reset the dynamic realm and matrix state at the same time (when the matrix state gets out of sync with the DB state chaos ensues, e.g. realms without owners or owners without realms)

dynamic_realm_creation

I think there needs to be more work on the dashboard around the interstitial/loading state for new realms. Filed as https://linear.app/cardstack/issue/CS-7235/workspace-chooser-loading-state

### CODE REVIEW

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (backspace) [inline packages/host/app/services/matrix-sdk-loader.ts]: What context are these used in? I’m sure I’m missing a part in my mental model because the Matrix client has this built in, right?

**Review by backspace — APPROVED** [review]
It took a few tries because I had lingering database connexions that made the `drop-all-dbs` script only partially work, but then it did finally 🎉 

Boxel 2024-09-20 17-13-02

I found when I logged out and logged in with the ≈original user that the new workspace was showing as well, unsure whether this is really related to this PR though, needs further investigation:

Boxel 2024-09-20 17-14-32

When I logged in via an incognito window, that workspace wasn’t there.

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (habdelra) [reply packages/host/app/services/matrix-sdk-loader.ts]: This is in the host app, the matrix js SDK is missing this (or at least our version of it is)

**Comment by habdelra** [comment]
> It took a few tries because I had lingering database connexions that made the `drop-all-dbs` script only partially work, but then it did finally 🎉
> 
> Boxel 2024-09-20 17-13-02
> I found when I logged out and logged in with the ≈original user that the new workspace was showing as well, unsure whether this is really related to this PR though, needs further investigation:
> 
> Boxel 2024-09-20 17-14-32
> When I logged in via an incognito window, that workspace wasn’t there.

Yeah, probably it’s not related to my changes, but I can dig into it regardless.

File: packages/realm-server/scripts/drop-all-dbs.sh
Comment (jurgenwerk) [inline packages/realm-server/scripts/drop-all-dbs.sh:22]: I love to see the added commands so that drop-all-dbs now takes care of everything

File: packages/host/app/lib/utils.ts
Comment (jurgenwerk) [inline packages/host/app/lib/utils.ts:38]: These icons look like they somehow diverge from the general boxel style, is this intended or is there a plan to replace them?
image

**Review by jurgenwerk — APPROVED** [review]
I tested this by first using `pnpm drop-all-dbs` and got the following crash when trying to run the app:

telegram-cloud-photo-size-4-6019280233531819384-y

I think it's probably unrelated to this PR and we are probably missing a handle on this, I think it should redirect to login when this happens.

After clearing my storage it worked and I was able to create my personal workspace 🎉

**Comment by habdelra** [comment]
> I tested this by first using `pnpm drop-all-dbs` and got the following crash when trying to run the app:
> 
> telegram-cloud-photo-size-4-6019280233531819384-y
> 
> I think it's probably unrelated to this PR and we are probably missing a handle on this, I think it should redirect to login when this happens.
> 
> After clearing my storage it worked and I was able to create my personal workspace 🎉

This error is because drop-all-dbs also clears your synapse server state, so your matrix auth token in localstorage now becomes invalid as synapse itself is regenerated with a new secret.

File: packages/host/app/lib/utils.ts
Comment (habdelra) [reply packages/host/app/lib/utils.ts:38]: These came from @lukemelia, he’ll have to chime in

**Comment by habdelra** [comment]
@backspace i tracked down the issue with seeing other people's workspaces to not clearing the card-service state when logging out.

File: packages/host/app/lib/utils.ts
Comment (lukemelia) [reply packages/host/app/lib/utils.ts:38]: They are a temporary solution made by me. We'll get Gardella to design something more in keeping with the rest of the system.

**Review by burieberry — APPROVED** [review]
Managed to get this to work after a couple of tries as well

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (lukemelia) [reply packages/host/app/services/matrix-sdk-loader.ts]: The matrix client has `getAccountDataFromServer`, which I belive does this.

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (lukemelia) [inline packages/host/app/services/matrix-sdk-loader.ts]: This will shadow the actual method with this name on matrix-js-sdk's client: https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/client.ts#L4232

File: packages/host/app/services/realm-server.ts
Comment (lukemelia) [inline packages/host/app/services/realm-server.ts:54]: Why are we using the loader's fetch here?

File: packages/host/app/services/realm-server.ts
Comment (lukemelia) [inline packages/host/app/services/realm-server.ts:168]: Same question about using the loader's fetch here. Seems like we should be using the virtual network's fetch.

**Review by lukemelia — DISMISSED** [review]

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (habdelra) [reply packages/host/app/services/matrix-sdk-loader.ts]: I added this, because despite being documented, it doesn't actually type check (so i just assumed it was missing). when i run it without these methods added in our code, despite not being typed it does work. so i guess we can just keep the types that I added but forgo our implementation

image

File: packages/host/app/services/realm-server.ts
Comment (habdelra) [reply packages/host/app/services/realm-server.ts:168]: is that because we want to bypass the middleware stack that the loader adds? i wasn't aware that we had to make those sorts of determinations outside of the loader service. doesn't our fetch abstraction leak when callers of the loader service need to make this kind of determination?

File: packages/host/app/services/realm-server.ts
Comment (habdelra) [reply packages/host/app/services/realm-server.ts:54]: same response as below. i can add this but it feels leaky...

File: packages/host/app/services/realm-server.ts
Comment (habdelra) [reply packages/host/app/services/realm-server.ts:168]: And followup, if there are circumstances that we do need to use the virutual network's fetch, what are the heuristics we can use to  know when to do that? and can we actually make a getter with a reasonable name to the virtual network's fetch that better informs callers of the loader service when they should reach for that instead of the loader's fetch (and what would that name be)?

keep in mind that the loader’s fetch is public (maybe it should not be), so that means there are 2 different fetches to choose from here. (and as an aside there are about 20 other places where we use the loader's fetch function in the host that should perhaps be revisited)

File: packages/host/app/services/matrix-service.ts
Comment (habdelra) [inline packages/host/app/services/matrix-service.ts:237]: @jurgenwerk just FYI, this is a minor collision with your PR to replace the realm server home screen. If my PR gets merged first you'll need to resolve this. which i think maybe just means deleting this altogether since the workspace chooser will be displayed after you first login via user registration.

Comment (jurgenwerk) [reply packages/host/app/services/matrix-service.ts:237]: Noted!

**Comment by habdelra** [comment]
I just tried this our in staging as I was testing the deploy and it took 12 minutes to create a new realm! something is amiss in the hosted envs.... I'm going to get to the bottom of this before I merge

File: packages/host/app/services/realm-server.ts
Comment (lukemelia) [reply packages/host/app/services/realm-server.ts:168]: I'll make a PR proposing some clarity around this. Good to keep this as is for now

**Comment by habdelra** [comment]
So the matrix rate limiting on logins (which is rate limited by IP) is causing havok with the dynamic realm indexing. The worker keeps triggering the login rate limit which either makes the indexing super slow and/or locks up the queue. We should not merge this until https://github.com/cardstack/boxel/pull/1596 is merged AND we have relaxed the synapse login rate limits.

File: packages/host/app/services/matrix-sdk-loader.ts
Comment (habdelra) [reply packages/host/app/services/matrix-sdk-loader.ts]: @lukemelia this is fixed now

**Pull request merged**
